From: Daniel van Vugt <daniel.van.vugt@canonical.com>
Date: Thu, 5 Mar 2020 11:34:23 +0100
Subject: overview: Fade in/out over the desktop instead of replacing it

This way extensions that add actors onto the background will fade
out/in smoothly from the overview.

Bug-Ubuntu: https://bugs.launchpad.net/bugs/1847712
Origin: https://gitlab.gnome.org/GNOME/gnome-shell/-/merge_requests/1070
---
 js/ui/layout.js   | 12 ++++++++++--
 js/ui/overview.js | 19 +++++++++----------
 2 files changed, 19 insertions(+), 12 deletions(-)

diff --git a/js/ui/layout.js b/js/ui/layout.js
index b958291..1eb4fd6 100644
--- a/js/ui/layout.js
+++ b/js/ui/layout.js
@@ -935,11 +935,19 @@ var LayoutManager = GObject.registerClass({
                                     monitor.inFullscreen);
     }
 
+    _setWindowsVisible(group, visible) {
+        for (let c = group.get_first_child(); c; c = c.get_next_sibling()) {
+            // We want to skip BackgroundActors which are also children...
+            if (c instanceof Meta.WindowActor)
+                c.visible = visible;
+        }
+    }
+
     _updateVisibility() {
         let windowsVisible = Main.sessionMode.hasWindows && !this._inOverview;
 
-        global.window_group.visible = windowsVisible;
-        global.top_window_group.visible = windowsVisible;
+        this._setWindowsVisible(global.window_group, windowsVisible);
+        this._setWindowsVisible(global.top_window_group, windowsVisible);
 
         this._trackedActors.forEach(this._updateActorVisibility.bind(this));
     }
diff --git a/js/ui/overview.js b/js/ui/overview.js
index 2e6b93d..ae8473b 100644
--- a/js/ui/overview.js
+++ b/js/ui/overview.js
@@ -244,11 +244,10 @@ var Overview = class {
     _unshadeBackgrounds() {
         let backgrounds = this._backgroundGroup.get_children();
         for (let i = 0; i < backgrounds.length; i++) {
-            backgrounds[i].ease_property('brightness', 1.0, {
-                duration: SHADE_ANIMATION_TIME,
-                mode: Clutter.AnimationMode.EASE_OUT_QUAD,
-            });
-            backgrounds[i].ease_property('vignette-sharpness', 0.0, {
+            backgrounds[i].ease({
+                brightness: 1.0,
+                vignette_sharpness: 0.0,
+                opacity: 0,
                 duration: SHADE_ANIMATION_TIME,
                 mode: Clutter.AnimationMode.EASE_OUT_QUAD,
             });
@@ -258,11 +257,11 @@ var Overview = class {
     _shadeBackgrounds() {
         let backgrounds = this._backgroundGroup.get_children();
         for (let i = 0; i < backgrounds.length; i++) {
-            backgrounds[i].ease_property('brightness', Lightbox.VIGNETTE_BRIGHTNESS, {
-                duration: SHADE_ANIMATION_TIME,
-                mode: Clutter.AnimationMode.EASE_OUT_QUAD,
-            });
-            backgrounds[i].ease_property('vignette-sharpness', Lightbox.VIGNETTE_SHARPNESS, {
+            backgrounds[i].set_opacity(0);
+            backgrounds[i].ease({
+                opacity: 255,
+                brightness: Lightbox.VIGNETTE_BRIGHTNESS,
+                vignette_sharpness: Lightbox.VIGNETTE_SHARPNESS,
                 duration: SHADE_ANIMATION_TIME,
                 mode: Clutter.AnimationMode.EASE_OUT_QUAD,
             });
